plugins/remark-lint-validate-internal-links/index.js

```diff
@@ -0,0 +1,204 @@
+      if (hasUpperCase(ref.split('/').pop())) {
+        file.message(warnings.pathWithUpperCase, node);
+        cache[ref] = warnings.pathWithUpperCase;
+        return;
+      }
+      addWarningIfFileIsMissing({
+        file,
+        node,
+        refRoot: staticRoot,
+        ref,
+        warning: warnings.missingStaticFile
+      }) || ( cache[ref] = warnings.valid );
+      return;
+    }
+    if (isDocRef(ref)) {
+      // paths with upper cases are not working
+      if (hasUpperCase(ref)) {
+        file.message(warnings.pathWithUpperCase, node);
+        cache[ref] = warnings.pathWithUpperCase;
+        return;
+      }
+      addWarningIfFileIsMissing({
+          file,
+          node,
+          refRoot: docsRoot,
+          ref: ref.endsWith('/') ? `${ref.slice(0, -1)}.md` : `${ref}.md`,
+          originalRef: ref,
+          warning: warnings.missingDoc
+      }) || ( cache[ref] = warnings.valid );
+      return;
+    }
+    verifyAnchor(file, node, ref);
+  }
+  
+  function verifyImages(node) {
+    const ref = node.url;
+    if (hasUpperCase(ref.split('/').pop())) {
+      file.message(warnings.pathWithUpperCase, node);
+      cache[ref] = warnings.pathWithUpperCase;
+      return;
+    }
+    addWarningIfFileIsMissing({
+      file,
+      node,
+      refRoot: staticRoot,
+      ref,
+      warning: warnings.missingImage
+    }) || ( cache[ref] = warnings.valid );
+  }  
+}
+
+function verifyAnchor(file, node, ref) {
+  let [filePath, anchor] = ref.split('#');
+  if (filePath.endsWith('/')) {
+    filePath = filePath.slice(0, -1)
+  }
+  filePath = join(projectRoot, docsRoot, `${filePath}.md`);
+  const doc = getFile({ file, node, ref, filePath });
+  if (doc) {
+    let anchorFound = 0;
+    // We got the doc now turn it into an AST.
+    const tree = fromMarkdown(doc);
+    visit(tree, "heading", compare);
+    function compare(currentNode) {
+      // Here we get the current node from AST which holds the heading value but
+      // it is the original one with spaces and upper case in the beggining.
+      // The references are with lower cases only and dashes instead of spaces
+      // to be URL friendly. So we do transform what we get as heading value in
+      // order to do proper comparison.
+      const refAnchor
+        = currentNode.children[0].value.toLowerCase().replace(' ', '-');
+      if (refAnchor === anchor) {
+        anchorFound++
+      }
+    }
+    if (anchorFound) {
+      cache[ref] = warnings.valid;
+    } else {
+      file.message(warnings.missingAnchor, node);
+      cache[ref] = warnings.missingAnchor;
+    }
+  }
+}
+
+function isExternalPageRef(ref) {
+  return ref.startsWith('http');
+}
+
+function isDocRef(ref) {
+  return !/#/g.test(ref);
+}
+
+function hasUpperCase(ref) {
+  return /[A-Z]/g.test(ref);
+}
+
+function isRelativeRef(ref) {
+  return ref.startsWith('../');
+}
+
+function isLocalAnchorRef(ref) {
+  return ref.startsWith('#');
+}
+
+function isStaticRef(ref) {
```

Comment:
**suggestion**: this is inefficient, calling `ref.endsWith` for each item in the array, and using reduce too.  a `path.extname`
```suggestion
function isStaticRef(ref) {
  const staticFileExtensions = ['.json', '.js', '.yaml', '.sh', '.html'];
  return staticFileExtensions.includes(path.extname(ref));
}
```